ptf/sai_base_test.py

```diff
@@ -0,0 +1,945 @@
+                            continue
+                        data[titles[i]] = item
+                    data['lanes'] = [int(lane)
+                                     for lane in data['lanes'].split(',')]
+                    data['speed'] = int(data['speed'])
+                    ports[name] = data
+            return ports
+        except Exception as e:
+            raise e
+
+    def checkPortsUp(self, timeout=30):
+        """
+        Wait for all ports to be UP
+        This may be required while testing on hardware
+        The test fails if all ports are not UP after timeout
+
+        Args:
+            timeout (int): port verification timeout in sec
+        """
+        allup = False
+        timer_start = time.time()
+
+        while allup is False and time.time() - timer_start < timeout:
+            allup = True
+            for port in self.port_list:
+                attr = sai_thrift_get_port_attribute(
+                    self.client, port, oper_status=True)
+                if attr['oper_status'] != SAI_SWITCH_OPER_STATUS_UP:
+                    allup = False
+                    break
+            if allup:
+                break
+            time.sleep(5)
+
+        self.assertTrue(allup)
+
+    def getSwitchPorts(self):
+        """
+        Get device port numbers
+        """
+        dev_no = 0
+        for _, port, _ in config['interfaces']:
+            setattr(self, 'dev_port%d' % dev_no, port)
+            dev_no += 1
+
+    def printNumberOfAvaiableResources(self, resources_dict):
+        """
+        Prints numbers of available resources
+
+        Args:
+            resources_dict (dict): a dictionary with resources numbers
+        """
+
+        print("***** Number of available resources *****")
+        for key, value in resources_dict:
+            print(key, ": ", value)
+
+    def saveNumberOfAvaiableResources(self, debug=False):
+        """
+        Save number of available resources
+        This allows to verify if all the test objects were removed
+
+        Args:
+            debug (bool): enables debug option
+        Return:
+            dict: switch_resources dictionary with available resources
+        """
+
+        switch_resources = sai_thrift_get_switch_attribute(
+            self.client,
+            available_ipv4_route_entry=True,
+            available_ipv6_route_entry=True,
+            available_ipv4_nexthop_entry=True,
+            available_ipv6_nexthop_entry=True,
+            available_ipv4_neighbor_entry=True,
+            available_ipv6_neighbor_entry=True,
+            available_next_hop_group_entry=True,
+            available_next_hop_group_member_entry=True,
+            available_fdb_entry=True,
+            available_ipmc_entry=True,
+            available_snat_entry=True,
+            available_dnat_entry=True,
+            available_double_nat_entry=True,
+            number_of_ecmp_groups=True,
+            ecmp_members=True)
```

Comment:
this is actually wrong approach, since not a lot of vendors has those implemented, for example current DASH implementation, and this function will throw sai thrift exception if any of those values will not be success, and this will then generate SkipTest exception if error is SAI_STATUS_NOT_SUPPORTED, which is what happens, and many test will/can be skipped if those values are not present, but they not prevent actually for test to be executed, each test should handle this independently or this function here should handle this in a way that if not supported, then set this resource to zero

Reply:
@lguohan @chrispsommers fyi

Reply:
@kcudnik  It does seem like the PTF infrastructure makes some assumptions about base capabilities of any SAI target which has caused some issues in DASH and forced some workarounds. It's arguable whether this is a flaw in the PTF framework and its assumptions, or that the DASH behavioral model falls too short of the mark for emulating a target in all its SAI support. Recent work by some vendors in SAI made many allowances for DASH such as supporting only 2-port devices, etc. but it didn't try to accommodate all the shortcomings if the current DASH behavioral model.

If there is opportunity to establish the capabilities/resources of the target and use that information to skip tests, that would be preferred.